# 3rd Year Bioinformatics practical

## Aims

    1 To develop the practical skills required to effectively implement bioinformatics workflows and data analysis. 
	2 To develop understanding of how to work collaboratively on data analysis 
	3 to understand provenance and sustainability in bioinformatics analysis
	
## Structure

The practical will consist of two parts. The first part is a 3 day 'boot camp' where you will be introduced to the techniques and concepts that you will be using. This is a very rapid 'crash course' and should be supplemented with your own reading and practice. You will not be expected to know everything you need for part 2, but you will have been introduced to the concepts and know where to find out the details. The second part is an exercise where you will put these skills into practice in your own time. 7 full working days have been allocated for the second part, and you can complete this at your own schedule.

## Part 1: Bootcamp

### Day 1

    * Setup and Introduction
	* Using Unix  
	* Python revision - variables
	* Python methods
	* Python classes.
	
### Day 2 
    * Documentation and testing
	* Data and code sharing with git
	* Relatinal databases (MySQL)
	* Python and relational databases

### Day 3
    * HTML and CSS
	* CGI forms
	* Python and CGI
	* Piecing it all together - Model, View, Controller
	
## Part 2: Exercise


You will download a gene expression dataset from a public database (eg GEO). The choice of dataset is up to you but should be confirmed with Dr Martin.  You will construct a suitable database structure to hold it, and provide a web interface to allow a user to interrogate that data (filter/select specific data by criteria you determine) which will then be presented as a web page. Your code should be fully documented and in revision control. You should show good separation between data structures (model), Data representation to the user (view) and the program code that controls this (controller).  